File: frontend/src/logic/actor.ts
import { Agent, Profession, Silo } from './models';

// ============ Actor 抽象 (文档 M0：状态归一) ============
/**
 * 统一 Actor 模型：
 * - PLAYER：特工 Agent (由 UI 控制)
 * - NPC：部门 Profession (由 NpcBrain 决策层控制)
 *
 * 执行引擎只认识 ActorView (统一状态视图)，不区分来源 ——
 * 玩家与 NPC 通过同一入口 submitAction 提交动作，共用同一套 AP/怀疑度/成本体系。
 */

export type ActorKind = 'PLAYER' | 'NPC';

export interface ActorRef {
  kind: ActorKind;
  /** PLAYER：特工 id */
  agentId?: number;
  /** 被控制的部门 id (NPC 必填；PLAYER 由 profession 推导) */
  professionId?: number;
}

export function createActorRefForAgent(agent: Agent, silo: Silo): ActorRef {
  const prof = silo.professions?.find(p => p.name === agent.profession);
  return { kind: 'PLAYER', agentId: agent.id, professionId: prof?.id };
}

export function createActorRefForProfession(prof: Profession): ActorRef {
  return { kind: 'NPC', professionId: prof.id };
}

export function createActorView(ref: ActorRef, silo: Silo, agent?: Agent): ActorView {
  if (ref.kind === 'PLAYER') {
    if (!agent) throw new Error('PLAYER actor requires agent reference.');
    return new ActorView(ref, agent, null, silo);
  }
  const prof = silo.professions?.find(p => p.id === ref.professionId);
  if (!prof) throw new Error(`NPC actor profession not found: ${ref.professionId}`);
  return new ActorView(ref, null, prof, silo);
}

/**
 * 统一状态视图：对 Agent / Profession 提供同构读写接口。
 * 执行管线 (engine) 与决策层 (NpcBrain) 均只依赖该视图。
 */
export class ActorView {
  constructor(
    public readonly ref: ActorRef,
    private readonly agent: Agent | null,
    private readonly prof: Profession | null,
    private readonly silo: Silo,
  ) {}

  get isPlayer(): boolean {
    return this.ref.kind === 'PLAYER';
  }

  /** 所属部门名称 (Agent.profession / Profession.name) */
  get profession(): string {
    return this.agent ? this.agent.profession : this.prof!.name;
  }

  /** 日志称谓：玩家显示特工名，NPC 显示部门名 */
  get label(): string {
    if (this.agent) return this.agent.name || this.agent.profession;
    return this.prof!.name;
  }

  // ============ 统一经济字段 (读写穿透到 Agent / Profession) ============

  get actionPoints(): number {
    return this.agent ? this.agent.action_points : (this.prof!.action_points ?? 0);
  }
  set actionPoints(v: number) {
    if (this.agent) this.agent.action_points = v;
    else this.prof!.action_points = v;
  }

  get suspicionLevel(): number {
    return this.agent ? this.agent.suspicion_level : (this.prof!.suspicion_level ?? 0);
  }
  set suspicionLevel(v: number) {
    if (this.agent) this.agent.suspicion_level = v;
    else this.prof!.suspicion_level = v;
  }

  get politicalPrestige(): number {
    return this.agent ? this.agent.political_prestige : (this.prof!.political_prestige ?? 0);
  }
  set politicalPrestige(v: number) {
    if (this.agent) this.agent.political_prestige = v;
    else this.prof!.political_prestige = v;
  }

  get propagandaLevel(): number {
    return this.agent ? this.agent.propaganda_level : (this.prof!.propaganda_level ?? 0);
  }
  set propagandaLevel(v: number) {
    if (this.agent) this.agent.propaganda_level = v;
    else this.prof!.propaganda_level = v;
  }

  get organizationFactor(): number {
    return this.agent ? this.agent.organization_factor : (this.prof!.organization_factor ?? 1.0);
  }
  set organizationFactor(v: number) {
    if (this.agent) this.agent.organization_factor = v;
    else this.prof!.organization_factor = v;
  }

  /** 政治点数：仅玩家特工拥有 (NPC 恒为 0) */
  get politicalPoints(): number {
    return this.agent ? this.agent.political_points : 0;
  }
  set politicalPoints(v: number) {
    if (this.agent) this.agent.political_points = v;
  }

  get traits(): string[] {
    if (this.agent) {
      if (!this.agent.traits) this.agent.traits = [];
      return this.agent.traits;
    }
    if (!this.prof!.traits) this.prof!.traits = [];
    return this.prof!.traits;
  }

  get knownFragments(): string[] {
    if (this.agent) {
      if (!this.agent.known_fragments) this.agent.known_fragments = [];
      return this.agent.known_fragments;
    }
    if (!this.prof!.known_fragments) this.prof!.known_fragments = [];
    return this.prof!.known_fragments;
  }

  // ============ 人脉 (统一按部门 id 访问) ============

  /** 该 Actor 对各地堡部门的人脉值数组 (用于平均威望计算) */
  connectionValues(): number[] {
    if (this.agent) {
      return (this.agent.connections ?? []).map(c => c.value);
    }
    return Object.values(this.prof!.relations ?? {});
  }

  getConnection(professionId: number): number {
    if (this.agent) {
      const c = this.agent.connections?.find(x => x.profession_id === professionId);
      return c ? c.value : 0;
    }
    const p = this.silo.professions?.find(x => x.id === professionId);
    return p ? (this.prof!.relations?.[p.name] ?? 0) : 0;
  }

  setConnection(professionId: number, value: number): void {
    if (this.agent) {
      if (!this.agent.connections) this.agent.connections = [];
      let c = this.agent.connections.find(x => x.profession_id === professionId);
      if (!c) {
        c = { id: Date.now(), agent_id: this.agent.id, profession_id: professionId, value: 0 };
        this.agent.connections.push(c);
      }
      c.value = value;
      return;
    }
    const p = this.silo.professions?.find(x => x.id === professionId);
    if (!p) return;
    if (!this.prof!.relations) this.prof!.relations = {};
    this.prof!.relations[p.name] = value;
  }

  addConnection(professionId: number, delta: number): void {
    this.setConnection(professionId, Math.min(1.0, Math.max(0, this.getConnection(professionId) + delta)));
  }
}

File: internal/engine/actor.go
package engine

import (
	"fmt"

	"silo40/internal/model"
)

// ============ Actor 抽象 (状态归一) ============
// 统一 Actor 模型：
// - PLAYER：特工 Agent (由 UI 控制)
// - NPC：部门 Profession (由 NpcBrain 决策层控制)
// 执行引擎只认识 ActorView (统一状态视图)，不区分来源。

type ActorKind string

const (
	ActorPlayer   ActorKind = "PLAYER"
	ActorNPC      ActorKind = "NPC"
	ActorResident ActorKind = "RESIDENT"
)

type ActorRef struct {
	Kind         ActorKind
	AgentID      uint // PLAYER：特工 id
	ProfessionID uint // 被控制的部门 id (NPC 必填；PLAYER 由 profession 推导)
	ResidentID   uint // RESIDENT：居民 id
}

func CreateActorRefForAgent(agent *model.Agent, silo *model.Silo) ActorRef {
	ref := ActorRef{Kind: ActorPlayer, AgentID: agent.ID}
	for _, p := range silo.Professions {
		if p.Name == agent.Profession {
			ref.ProfessionID = p.ID
			break
		}
	}
	return ref
}

func CreateActorRefForProfession(prof *model.Profession) ActorRef {
	return ActorRef{Kind: ActorNPC, ProfessionID: prof.ID}
}

func CreateActorRefForResident(resident *model.Resident) ActorRef {
	return ActorRef{Kind: ActorResident, ResidentID: resident.ID, ProfessionID: resident.ProfessionID}
}

// ActorView 统一状态视图：对 Agent / Profession 提供同构读写接口
type ActorView struct {
	Ref      ActorRef
	agent    *model.Agent
	prof     *model.Profession
	resident *model.Resident
	silo     *model.Silo
}

func CreateActorView(ref ActorRef, silo *model.Silo, agent *model.Agent) (*ActorView, error) {
	if ref.Kind == ActorPlayer {
		if agent == nil {
			return nil, fmt.Errorf("PLAYER actor requires agent reference")
		}
		return &ActorView{Ref: ref, agent: agent, silo: silo}, nil
	}
	if ref.Kind == ActorResident {
		for i := range silo.Residents {
			if silo.Residents[i].ID == ref.ResidentID {
				return &ActorView{Ref: ref, resident: &silo.Residents[i], silo: silo}, nil
			}
		}
		return nil, fmt.Errorf("resident actor not found: %d", ref.ResidentID)
	}
	for i := range silo.Professions {
		if silo.Professions[i].ID == ref.ProfessionID {
			return &ActorView{Ref: ref, prof: &silo.Professions[i], silo: silo}, nil
		}
	}
	return nil, fmt.Errorf("NPC actor profession not found: %d", ref.ProfessionID)
}

func (v *ActorView) IsPlayer() bool { return v.Ref.Kind == ActorPlayer }

// Profession 所属部门名称
func (v *ActorView) Profession() string {
	if v.agent != nil {
		return v.agent.Profession
	}
	if v.resident != nil {
		return v.resident.Profession
	}
	return v.prof.Name
}

// Label 日志称谓：玩家显示特工名，NPC 显示部门名
func (v *ActorView) Label() string {
	if v.agent != nil {
		if v.agent.Name != "" {
			return v.agent.Name
		}
		return v.agent.Profession
	}
	if v.resident != nil {
		if v.resident.Name != "" {
			return v.resident.Name
		}
		return v.resident.Profession
	}
	return v.prof.Name
}

// ============ 统一经济字段 (读写穿透到 Agent / Profession) ============

func (v *ActorView) ActionPoints() float64 {
	if v.agent != nil {
		return v.agent.ActionPoints
	}
	if v.resident != nil {
		return v.resident.ActionPoints
	}
	return v.prof.ActionPoints
}

func (v *ActorView) SetActionPoints(val float64) {
	if v.agent != nil {
		v.agent.ActionPoints = val
	} else if v.resident != nil {
		v.resident.ActionPoints = val
	} else {
		v.prof.ActionPoints = val
	}
}

func (v *ActorView) SuspicionLevel() float64 {
	if v.agent != nil {
		return v.agent.SuspicionLevel
	}
	if v.resident != nil {
		return v.resident.SuspicionLevel
	}
	return v.prof.SuspicionLevel
}

func (v *ActorView) SetSuspicionLevel(val float64) {
	if v.agent != nil {
		v.agent.SuspicionLevel = val
	} else if v.resident != nil {
		v.resident.SuspicionLevel = val
	} else {
		v.prof.SuspicionLevel = val
	}
}

func (v *ActorView) PoliticalPrestige() float64 {
	if v.agent != nil {
		return v.agent.PoliticalPrestige
	}
	if v.resident != nil {
		return v.resident.PoliticalPrestige
	}
	return v.prof.PoliticalPrestige
}

func (v *ActorView) SetPoliticalPrestige(val float64) {
	if v.agent != nil {
		v.agent.PoliticalPrestige = val
	} else if v.resident != nil {
		v.resident.PoliticalPrestige = val
	} else {
		v.prof.PoliticalPrestige = val
	}
}

func (v *ActorView) PropagandaLevel() float64 {
	if v.agent != nil {
		return v.agent.PropagandaLevel
	}
	if v.resident != nil {
		return v.resident.PropagandaLevel
	}
	return v.prof.PropagandaLevel
}

func (v *ActorView) SetPropagandaLevel(val float64) {
	if v.agent != nil {
		v.agent.PropagandaLevel = val
	} else if v.resident != nil {
		v.resident.PropagandaLevel = val
	} else {
		v.prof.PropagandaLevel = val
	}
}

// Productivity 生产力：读写 Actor 所属部门的生产力
func (v *ActorView) Productivity() float64 {
	if v.prof != nil {
		return v.prof.Productivity
	}
	if v.resident != nil {
		for i := range v.silo.Professions {
			if v.silo.Professions[i].ID == v.resident.ProfessionID {
				return v.silo.Professions[i].Productivity
			}
		}
	}
	for i := range v.silo.Professions {
		if v.silo.Professions[i].Name == v.agent.Profession {
			return v.silo.Professions[i].Productivity
		}
	}
	return 1
}

func (v *ActorView) SetProductivity(val float64) {
	if v.prof != nil {
		v.prof.Productivity = val
		return
	}
	if v.resident != nil {
		for i := range v.silo.Professions {
			if v.silo.Professions[i].ID == v.resident.ProfessionID {
				v.silo.Professions[i].Productivity = val
				return
			}
		}
	}
	for i := range v.silo.Professions {
		if v.silo.Professions[i].Name == v.agent.Profession {
			v.silo.Professions[i].Productivity = val
			return
		}
	}
}

// PoliticalPoints 政治点数：仅玩家特工拥有 (NPC 恒为 0)
func (v *ActorView) PoliticalPoints() float64 {
	if v.agent != nil {
		return v.agent.PoliticalPoints
	}
	return 0
}

func (v *ActorView) SetPoliticalPoints(val float64) {
	if v.agent != nil {
		v.agent.PoliticalPoints = val
	}
}

func (v *ActorView) Traits() []string {
	if v.agent != nil {
		if v.agent.Traits == nil {
			v.agent.Traits = []string{}
		}
		return v.agent.Traits
	}
	if v.resident != nil {
		if v.resident.Tags == nil {
			v.resident.Tags = []string{}
		}
		return v.resident.Tags
	}
	if v.prof.Traits == nil {
		v.prof.Traits = []string{}
	}
	return v.prof.Traits
}

func (v *ActorView) KnownFragments() []string {
	if v.agent != nil {
		if v.agent.KnownFragments == nil {
			v.agent.KnownFragments = []string{}
		}
		return v.agent.KnownFragments
	}
	if v.resident != nil {
		if v.resident.KnownFragments == nil {
			v.resident.KnownFragments = []string{}
		}
		return v.resident.KnownFragments
	}
	if v.prof.KnownFragments == nil {
		v.prof.KnownFragments = []string{}
	}
	return v.prof.KnownFragments
}

// ============ 人脉 (统一按部门 id 访问) ============

// ConnectionValues 该 Actor 对各地堡部门的人脉值数组 (用于平均威望计算)
func (v *ActorView) ConnectionValues() []float64 {
	if v.agent != nil {
		vals := make([]float64, 0, len(v.agent.Connections))
		for _, c := range v.agent.Connections {
			vals = append(vals, c.Value)
		}
		return vals
	}
	if v.resident != nil {
		vals := make([]float64, 0, len(v.resident.Relations))
		for _, val := range v.resident.Relations {
			vals = append(vals, val)
		}
		return vals
	}
	vals := make([]float64, 0, len(v.prof.Relations))
	for _, val := range v.prof.Relations {
		vals = append(vals, val)
	}
	return vals
}

func (v *ActorView) GetConnection(professionID uint) float64 {
	if v.agent != nil {
		for _, c := range v.agent.Connections {
			if c.ProfessionID == professionID {
				return c.Value
			}
		}
		return 0
	}
	if v.resident != nil {
		for i := range v.silo.Professions {
			if v.silo.Professions[i].ID == professionID {
				return v.resident.Relations[v.silo.Professions[i].Name]
			}
		}
		return 0
	}
	for i := range v.silo.Professions {
		if v.silo.Professions[i].ID == professionID {
			return v.prof.Relations[v.silo.Professions[i].Name]
		}
	}
	return 0
}

func (v *ActorView) SetConnection(professionID uint, value float64) {
	if v.agent != nil {
		if v.agent.Connections == nil {
			v.agent.Connections = []model.Connection{}
		}
		for i := range v.agent.Connections {
			if v.agent.Connections[i].ProfessionID == professionID {
				v.agent.Connections[i].Value = value
				return
			}
		}
		v.agent.Connections = append(v.agent.Connections, model.Connection{
			AgentID:      v.agent.ID,
			ProfessionID: professionID,
			Value:        value,
		})
		return
	}
	if v.resident != nil {
		for i := range v.silo.Professions {
			if v.silo.Professions[i].ID == professionID {
				if v.resident.Relations == nil {
					v.resident.Relations = map[string]float64{}
				}
				v.resident.Relations[v.silo.Professions[i].Name] = value
				return
			}
		}
		return
	}
	for i := range v.silo.Professions {
		if v.silo.Professions[i].ID == professionID {
			if v.prof.Relations == nil {
				v.prof.Relations = map[string]float64{}
			}
			v.prof.Relations[v.silo.Professions[i].Name] = value
			return
		}
	}
}

func (v *ActorView) AddConnection(professionID uint, delta float64) {
	cur := v.GetConnection(professionID) + delta
	if cur < 0 {
		cur = 0
	}
	if cur > 1.0 {
		cur = 1.0
	}
	v.SetConnection(professionID, cur)
}
